--- target/riscv/vector_helper.bkp	2026-03-04 15:40:55.548532273 +0100
+++ target/riscv/vector_helper.c	2026-03-04 16:03:56.373663753 +0100
@@ -381,8 +381,6 @@
     uint32_t nf = vext_nf(desc);
     uint32_t vlmax = vext_maxsz(desc) / esz;
 
-    /* probe every access */
-    probe_pages(env, base, env->vl * nf * msz, ra, access_type);
     /* load bytes from guest memory */
     for (i = 0; i < env->vl; i++) {
         k = 0;
